File: ios/KidCanvasApp/Views/StoryViews.swift
import SwiftUI

/// The story field is the point of the app: what the child said about the
/// artwork, captured before it's forgotten.
struct StoryField: View {
    @Binding var story: String
    let childName: String?
    let onBrowseTemplates: () -> Void

    private var remaining: Int {
        max(0, StoryRules.minimumLength - story.trimmed.count)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text("The story")
                    .font(.subheadline.bold())
                    .foregroundColor(.secondary)
                Spacer()
                Button(action: onBrowseTemplates) {
                    Label("Need a nudge?", systemImage: "lightbulb")
                        .font(.caption)
                }
                .accessibilityLabel("Browse story prompts")
            }

            TextField(prompt, text: $story, axis: .vertical)
                .lineLimit(4...8)
                .padding()
                .background(Color.cardSurface)
                .cornerRadius(12)
                .accessibilityLabel("The story behind this artwork")

            Text(remaining > 0
                 ? "A sentence or two — \(remaining) more character\(remaining == 1 ? "" : "s")"
                 : "This is what you'll be glad you wrote down.")
                .font(.caption2)
                .foregroundColor(remaining > 0 ? .secondary : .green)
        }
    }

    private var prompt: String {
        if let childName {
            return "What did \(childName) say about this?"
        }
        return "What did they say about this?"
    }
}

struct StoryTemplatePicker: View {
    let onPick: (String) -> Void

    @Environment(\.dismiss) private var dismiss
    @State private var category: StoryTemplate.Category = .milestone

    var body: some View {
        NavigationStack {
            VStack(spacing: 0) {
                CategoryStrip(selected: $category)
                List(StoryTemplates.templates(in: category)) { template in
                    Button {
                        onPick(template.opener)
                    } label: {
                        HStack(spacing: 12) {
                            Text(template.icon)
                                .font(.title3)
                            VStack(alignment: .leading, spacing: 2) {
                                Text(template.title)
                                    .font(.subheadline.weight(.semibold))
                                    .foregroundColor(.primary)
                                Text(template.opener + "…")
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                                    .lineLimit(2)
                            }
                        }
                    }
                }
                .listStyle(.plain)
            }
            .navigationTitle("Story prompts")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Cancel") { dismiss() }
                }
            }
        }
    }
}

struct CategoryStrip: View {
    @Binding var selected: StoryTemplate.Category

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(StoryTemplate.Category.allCases) { category in
                    Button {
                        selected = category
                    } label: {
                        Text(category.rawValue)
                            .font(.caption.weight(.medium))
                            .padding(.horizontal, 14)
                            .padding(.vertical, 8)
                            .background(
                                selected == category
                                    ? AnyShapeStyle(LinearGradient(colors: [.pink, .purple],
                                                                   startPoint: .leading,
                                                                   endPoint: .trailing))
                                    : AnyShapeStyle(Color.subtleFill),
                                in: Capsule()
                            )
                            .foregroundColor(selected == category ? .white : .primary)
                    }
                }
            }
            .padding(.horizontal)
            .padding(.vertical, 10)
        }
    }
}

/// The story as it appears on the artwork detail screen.
struct StoryCard: View {
    let story: String
    let childName: String?

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Label("The story", systemImage: "text.quote")
                .font(.subheadline.bold())
                .foregroundColor(.secondary)
            Text(story)
                .font(.body)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(16)
        .background(Color.pink.opacity(0.06), in: RoundedRectangle(cornerRadius: 16))
        .accessibilityElement(children: .combine)
        .accessibilityLabel("Story\(childName.map { " about \($0)'s artwork" } ?? ""): \(story)")
    }
}

/// Prompts the parent to add a story to artwork saved before the field existed.
struct AddStoryPrompt: View {
    let onAdd: () -> Void

    var body: some View {
        Button(action: onAdd) {
            HStack {
                Image(systemName: "text.badge.plus")
                    .foregroundColor(.pink)
                VStack(alignment: .leading, spacing: 2) {
                    Text("No story yet")
                        .font(.subheadline.weight(.semibold))
                        .foregroundColor(.primary)
                    Text("Add what they said about it")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
                Spacer()
                Image(systemName: "chevron.right")
                    .font(.caption)
                    .foregroundColor(.secondary)
            }
            .padding(16)
            .background(Color.faintFill, in: RoundedRectangle(cornerRadius: 16))
        }
        .buttonStyle(.plain)
    }
}

File: android/app/src/main/java/app/kidcanvas/ui/story/StoryViews.kt
package app.kidcanvas.ui.story

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.KeyboardArrowRight
import androidx.compose.material.icons.filled.Lightbulb
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.ui.unit.dp
import app.kidcanvas.data.StoryTemplate
import app.kidcanvas.data.StoryTemplates
import app.kidcanvas.ui.common.GradientButton
import kotlinx.coroutines.launch

/**
 * The story field is the point of the app: what the child said about the
 * artwork, captured before it is forgotten. Same copy and prompt as iOS.
 */
@Composable
fun StoryField(story: String, onStoryChange: (String) -> Unit, childName: String?) {
    var showTemplates by rememberSaveable { mutableStateOf(false) }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text("The story", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Bold,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            Spacer(Modifier.width(6.dp))
            Text("optional", style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            Spacer(Modifier.weight(1f))
            TextButton(onClick = { showTemplates = true }) {
                Icon(Icons.Filled.Lightbulb, contentDescription = null, modifier = Modifier.size(16.dp))
                Spacer(Modifier.width(4.dp))
                Text("Need a nudge?", style = MaterialTheme.typography.labelMedium)
            }
        }
        OutlinedTextField(
            value = story,
            onValueChange = onStoryChange,
            placeholder = { Text(childName?.let { "What did $it say about this?" } ?: "What did they say about this?") },
            minLines = 4,
            maxLines = 8,
            keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Sentences),
            modifier = Modifier.fillMaxWidth().semantics { contentDescription = "The story behind this artwork" },
        )
        Text(
            "A sentence or two. It's the part you'll be glad you wrote down.",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
    if (showTemplates) {
        StoryTemplatePicker(
            onPick = { opener ->
                onStoryChange(StoryTemplates.apply(opener, story))
                showTemplates = false
            },
            onDismiss = { showTemplates = false },
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun StoryTemplatePicker(onPick: (String) -> Unit, onDismiss: () -> Unit) {
    var category by rememberSaveable { mutableStateOf(StoryTemplate.Category.Milestone) }
    ModalBottomSheet(onDismissRequest = onDismiss) {
        Text("Story prompts", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold,
            modifier = Modifier.padding(horizontal = 20.dp))
        Row(
            Modifier.horizontalScroll(rememberScrollState()).padding(horizontal = 16.dp, vertical = 10.dp),
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            StoryTemplate.Category.entries.forEach { c ->
                FilterChip(selected = c == category, onClick = { category = c }, label = { Text(c.label) })
            }
        }
        LazyColumn(Modifier.fillMaxWidth().heightIn(max = 420.dp), contentPadding = PaddingValues(bottom = 24.dp)) {
            items(StoryTemplates.templates(category), key = { it.id }) { t ->
                Row(
                    Modifier.fillMaxWidth().clickable { onPick(t.opener) }.padding(horizontal = 20.dp, vertical = 12.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(t.icon, style = MaterialTheme.typography.titleLarge)
                    Spacer(Modifier.width(12.dp))
                    Column {
                        Text(t.title, style = MaterialTheme.typography.bodyLarge, fontWeight = FontWeight.SemiBold)
                        Text(t.opener.trimEnd() + "…", style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant, maxLines = 2)
                    }
                }
            }
        }
    }
}

/** The story as it appears on the detail screen: a tinted card, never a caption. */
@Composable
fun StoryCard(story: String, childName: String?) {
    Column(
        Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.primary.copy(alpha = 0.06f), RoundedCornerShape(16.dp))
            .padding(16.dp)
            .semantics(mergeDescendants = true) {
                contentDescription = "Story" + (childName?.let { " about $it's artwork" } ?: "") + ": $story"
            },
    ) {
        Text("The story", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Bold,
            color = MaterialTheme.colorScheme.onSurfaceVariant)
        Spacer(Modifier.height(8.dp))
        Text(story, style = MaterialTheme.typography.bodyLarge)
    }
}

/** For artwork saved without a story: the app has months of images with no words attached. */
@Composable
fun AddStoryPrompt(onAdd: () -> Unit) {
    Row(
        Modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.04f), RoundedCornerShape(16.dp))
            .clickable(onClick = onAdd)
            .padding(16.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text("No story yet", style = MaterialTheme.typography.bodyMedium, fontWeight = FontWeight.SemiBold)
            Text("Add what they said about it", style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
        Icon(Icons.AutoMirrored.Filled.KeyboardArrowRight, contentDescription = null)
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AddStorySheet(childName: String?, onSave: suspend (String) -> String?, onDismiss: () -> Unit) {
    var story by rememberSaveable { mutableStateOf("") }
    var saving by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)) {
        Column(Modifier.padding(horizontal = 20.dp).padding(bottom = 24.dp).imePadding(), verticalArrangement = Arrangement.spacedBy(16.dp)) {
            Text("Add the story", style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.SemiBold)
            StoryField(story, { story = it }, childName)
            errorText?.let { Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall) }
            GradientButton(
                text = "Save story",
                enabled = story.isNotBlank(),
                busy = saving,
                onClick = {
                    saving = true
                    scope.launch {
                        errorText = onSave(story.trim())
                        saving = false
                        if (errorText == null) onDismiss()
                    }
                },
            )
        }
    }
}
